File: graphs/breadth-first-search/bfs.js
/* A Queue object for queue-like functionality over JavaScript arrays. */
let Queue = function () {
    this.items = [];
};
Queue.prototype.enqueue = function (obj) {
    this.items.push(obj);
};
Queue.prototype.dequeue = function () {
    return this.items.shift();
};
Queue.prototype.isEmpty = function () {
    return this.items.length === 0;
};

/**
 * Performs a breadth-first search on a graph
 * @param {array} graph - Graph, represented as adjacency lists.
 * @param {number} source - The index of the source vertex.
 * @return {array} Array of objects describing each vertex, like
 *     [{distance: _, predecessor: _ }]
 */
let bfs = function (graph, source) {
    let vertexInfo = [];
    // Index of each object -> vertex number, v
    for (let i = 0; i < graph.length; i++) {
        vertexInfo[i] = {
            // Initialise to null to track which vertices have been processed
            distance: null,
            predecessor: null
        };
    }

    // Source vertex has no predecessor
    vertexInfo[source].distance = 0;

    let queue = new Queue();
    queue.enqueue(source);

    while (!queue.isEmpty()) {
        let u = queue.dequeue();
        // Search through neighbours of u
        for (let i = 0; i < graph[u].length; i++) {
            let v = graph[u][i];
            // Process if not already processed
            if (vertexInfo[v].distance === null) {
                // Update info for u
                vertexInfo[v].distance = vertexInfo[u].distance + 1;
                vertexInfo[v].predecessor = u;
                // Add to queue to process its neighbours later
                queue.enqueue(v);
            }
        }
    }

    return vertexInfo;
};


// Testing ------------------
/**
 * Compares property value level equality of two objects
 * @param {object} a 
 * @param {object} b
 * @return {boolean}
 */
let isObjectEquivalent = function (a, b) {
    // Create arrays of property names
    let aProps = Object.getOwnPropertyNames(a);
    let bProps = Object.getOwnPropertyNames(b);
    // Check number of properties
    // Compare equality of each property value
    return aProps.length === bProps.length &&
        aProps.reduce((prev, curr, i) => a[aProps[i]] === b[aProps[i]], true)
}

// Adjacency lists representing graph
let adjList = [
    [1],
    [0, 4, 5],
    [3, 4, 5],
    [2, 6],
    [1, 2],
    [1, 2, 6],
    [3, 5],
    []
];

let vertexInfo = bfs(adjList, 3);
for (let i = 0; i < adjList.length; i++) {
    console.log("vertex " + i + ": distance = " + vertexInfo[i].distance + ", predecessor = " + vertexInfo[i].predecessor);
}

console.assert(true === isObjectEquivalent(vertexInfo[0], {
    distance: 4,
    predecessor: 1
}));
console.assert(true === isObjectEquivalent(vertexInfo[1], {
    distance: 3,
    predecessor: 4
}));
console.assert(true === isObjectEquivalent(vertexInfo[2], {
    distance: 1,
    predecessor: 3
}));
console.assert(true === isObjectEquivalent(vertexInfo[3], {
    distance: 0,
    predecessor: null
}));
console.assert(true === isObjectEquivalent(vertexInfo[4], {
    distance: 2,
    predecessor: 2
}));
console.assert(true === isObjectEquivalent(vertexInfo[5], {
    distance: 2,
    predecessor: 2
}));
console.assert(true === isObjectEquivalent(vertexInfo[6], {
    distance: 1,
    predecessor: 3
}));
console.assert(true === isObjectEquivalent(vertexInfo[7], {
    distance: null,
    predecessor: null
}));

File: graphs/breadth-first-search/bfs.py
from collections import deque


def bfs(graph, source):
    """ Performs a breadth-first search on a graph

    Args:
        graph (list of list of int): Adjacency matrix representation of graph
        source (int): Index of source vertex to begin search from

    Returns:
        list of dicts describing each vertex -> [{distance: _, predecessor: _ }] 
    """
    vertexInfo = []
    for i in range(len(graph)):
        vertexInfo.append({"distance": None, "predecessor": None})
    vertexInfo[source]["distance"] = 0

    search_queue = deque()
    search_queue.append(source)

    while search_queue:
        u = search_queue.popleft()
        for v in graph[u]:
            if vertexInfo[v]["distance"] is None:
                vertexInfo[v]["distance"] = vertexInfo[u]["distance"] + 1
                vertexInfo[v]["predecessor"] = u
                search_queue.append(v)
    return vertexInfo


# Testing --------------------

adjList = [
    [1],
    [0, 4, 5],
    [3, 4, 5],
    [2, 6],
    [1, 2],
    [1, 2, 6],
    [3, 5],
    []
]

vertexInfo = bfs(adjList, 3)

for i in range(len(adjList)):
    print("vertex %s : distance = %s, predecessor = %s" %
          (i, vertexInfo[i]["distance"], vertexInfo[i]["predecessor"]));

assert(vertexInfo[0] == {
    "distance": 4,
    "predecessor": 1
});
assert(vertexInfo[1] == {
    "distance": 3,
    "predecessor": 4
});
assert(vertexInfo[2] == {
    "distance": 1,
    "predecessor": 3
});
assert(vertexInfo[3] == {
    "distance": 0,
    "predecessor": None
});
assert(vertexInfo[4] == {
    "distance": 2,
    "predecessor": 2
});
assert(vertexInfo[5] == {
    "distance": 2,
    "predecessor": 2
});
assert(vertexInfo[6] == {
    "distance": 1,
    "predecessor": 3
});
assert(vertexInfo[7] == {
    "distance": None,
    "predecessor": None
});
